### PR TITLE
Adopt structured concurrency in `forEach` functions

### PR DESCRIPTION
**Pull request by Joannis**
Gets rid of unstructured tasks, in favour of a _structured_ concurrency approach. In addition, this PR makes the APIs for cancelling changeStreams more explicit

### CODE REVIEW

File: Sources/MongoKitten/ChangeStream.swift [hunk 1]
```diff
@@ -139,25 +151,45 @@ public struct ChangeStream<T: Decodable> {
...
-                }
-                
-                if let getMoreInterval = self.getMoreInterval {
-                    try await Task.sleep(nanoseconds: UInt64(getMoreInterval.nanoseconds))
+            try await forEach { notification in
+                if try await handler(notification) {
+                    return .continue
                 } else {
-                    try Task.checkCancellation()
+                    return .stop
```
Comment (Obbut) [inline Sources/MongoKitten/ChangeStream.swift:166]: Isn't this a breaking change if somebody relies on catching cancellation errors?

Comment (Joannis) [reply Sources/MongoKitten/ChangeStream.swift:166]: This is still the case:

https://github.com/orlandos-nl/MongoKitten/pull/329/files#diff-120634bc6a29b714a4b68a7f15f5a4fe6a6ef678bbf0aa5abeafef9eed7ef0e2R191

File: Sources/MongoKitten/ChangeStream.swift [hunk 2]
```diff
@@ -139,25 +151,45 @@ public struct ChangeStream<T: Decodable> {
...
+            }
+        }
+    }
+
+    /// Iterates over the change stream notifications and calls the given handler for each notification
+    /// 
+    /// - Parameter handler: The handler to call for each notification
+    /// - Throws: When the handler throws an error or a network error occurs
+    /// - Returns: When the change stream ends, due to Task cancellation or a closed cursor
+    public func forEach(handler: @escaping @Sendable (Notification) async throws -> ChangeStreamPollingResult) async throws {
```
Comment (Obbut) [inline Sources/MongoKitten/ChangeStream.swift:177]: As discussed, we should double-check that the different closure types don't break any existing codebases.

**Review by Obbut — APPROVED** [review]

File: Sources/MongoKitten/ChangeStream.swift (hunk 1 above)
Comment (Obbut) [reply Sources/MongoKitten/ChangeStream.swift:166]: Right, the diff was a bit confusing. Thanks for clarifying 🙃

**Pull request closed without merging**